model/modules/model/parse_country_model_module.py

```diff
@@ -71,6 +84,66 @@ def _apply_append_after(self, model: Model, yaml: Any) -> None:
           parent.children = (parent.children[:insert_idx + 1] + nodes +
                              parent.children[insert_idx + 1:])
 
+  def _find_lowest_common_ancestor(self, model: Model,
+                                   children: List[str]) -> str:
+    """
+    Returns the lowest common ancestor of all children in the model tree.
+
+    Args:
+      model: The model
+      children: A list of token_ids from which we look for the lowest
+         common ancestor.
+    """
+
+    # Store for each child in children a path from the root node to the child.
+    # Each path is a list of token_ids. paths_to_node is a list of len(children)
+    # such paths.
+    paths_to_nodes = [
+        model.find_path_to_node(child_id) for child_id in children
+    ]
+
+    # Remove the first entries from all lists and store them in last_heads.
+    last_heads = [path.pop(0) for path in paths_to_nodes]
+
+    while True:
+      # At this point all heads should share the same node.
+      assert all(last_heads[0] == l for l in last_heads)
+
+      # What is the minimum length of all remaining paths?
+      min_remaining_length = min([len(path) for path in paths_to_nodes])
+
+      # At least the leaf nodes should remain
+      assert min_remaining_length >= 1
+
+      # If one remaining path starts starts with a different head, we have found
+      # the lowest common ancestor.
+      if not all(paths_to_nodes[0][0] == path[0] for path in paths_to_nodes):
+        return last_heads[0]
+
+      # Remove the first entries from all lists and store them in last_heads
+      # again.
+      last_heads = [path.pop(0) for path in paths_to_nodes]
+
+  def _apply_synthesized_nodes(self, model: Model, yaml: Any) -> None:
+    synthesized_nodes = yaml.get('synthesized-nodes', {})
+    for id, children in synthesized_nodes.items():
+      # Verify properties of synthesized nodes:
+      assert id not in model.concepts
+      assert len(children) > 0
+      for child_id in children:
+        assert child_id in model.concepts
+        child = model.concepts[child_id]
+        if not child.is_atomic_token():
+          assert not isinstance(child, AtomicToken)
+          assert not child.is_synthesized
+      new_token = CompoundToken(model, id, children, is_synthesized=True)
+      model.concepts[id] = new_token
+
+      lowest_common_anecstor = self._find_lowest_common_ancestor(
+          model, children)
+      assert lowest_common_anecstor
+      model.concepts[lowest_common_anecstor].children.append(id)
```

Comment:
So IIUC, sinthesized nodes will be included as part of the LCA's children? Was there anything any particular reason to set it up that way?

I ask to make sure whether it is necessary to do a similar thing in Chrome.

Reply:
I add the nodes so that they get rendered on https://battre.github.io/autocomplete-attribute-explainer/model/IN.html - you should probably treat them separately in Chrome's hierarchy.